File: merge_sort/merge_sort.go
package main

import "fmt"

// MergeSort divides the slice into two parts, sorts both the part of slice and then merge it
// Time complexity : O(size(log(size)))
// Space complexity : O(size)
func MergeSort(slice []int) []int {
	if len(slice) < 2 {
		return slice
	}
	mid := (len(slice)) / 2
	return merge(MergeSort(slice[:mid]), MergeSort(slice[mid:]))
}

// merge merges two sorted slice
// Time complexity : O(size)
// Space complexity : O(size)
func merge(left []int, right []int) []int {
	i := 0
	j := 0
	size := len(left) + len(right)
	slice := make([]int, size, size) // make built-in function allocates and initializes an object of type slice
	for k := 0; k < size; k++ {
		if i > len(left)-1 && j < len(right) {
			slice[k] = right[j]
			j++
		} else if j > len(right)-1 && i < len(left) {
			slice[k] = left[i]
			i++
		} else if left[i] < right[j] {
			slice[k] = left[i]
			i++
		} else {
			slice[k] = right[j]
			j++
		}
	}
	return slice
}

func main() {
	slice := []int{8, 2, 4, 9, 1, 699, 0, 9999}
	fmt.Println("Sorted data is:")
	slice = MergeSort(slice)
	fmt.Println(slice)
}

File: merge_sort/mergeSort.js
/*
*	Implementation of merge sort algorithm
*	Time complexity: O(size*log(size))
*	Space complexity: O(size)
*/
function merge (left, right) {
	let i = 0;
	let j = 0;
	let size = left.length + right.length;
	let arr = [];
	for (let k = 0; k < size; k++) {
		if (i > left.length - 1 && j < right.length) {
			arr[k] = right[j];
			j++;
		} else if (j > right.length - 1 && i < left.length) {
			arr[k] = left[i];
			i++;
		} else if (left[i] < right[j]) {
			arr[k] = left[i];
			i++;
		} else {
			arr[k] = right[j];
			j++;
		}
	}
	return arr;
}

/*
*	:param arr: array to be sorted
*/
function mergeSort (arr) {
	if (arr.length < 2) {
		return arr;
	}
	let middle = parseInt(arr.length / 2);
	let left = arr.slice(0, middle);
	let right = arr.slice(middle, arr.length);
	return merge(mergeSort(left), mergeSort(right));
}

function main () {
	let arr = [2, 1, 6, 444, 8, 99, 10];
	console.log('Sorted data is :');
	arr = mergeSort(arr);
	console.log(arr);
}

main();
